drivers/media/platform/raspberrypi/pisp_be/pisp_be.c

```diff
@@ -0,0 +1,1972 @@
+		    !(hw_enables[0] & PISP_BE_BAYER_ENABLE_TDN) ||
+		    (config->config.tdn.reset & 1)) {
+			hw_enables[0] &= ~(PISP_BE_BAYER_ENABLE_TDN_INPUT |
+					   PISP_BE_BAYER_ENABLE_TDN_DECOMPRESS);
+			if (!(config->config.tdn.reset & 1))
+				hw_enables[0] &= ~PISP_BE_BAYER_ENABLE_TDN;
+		}
+
+		addrs[4] = pispbe_get_addr(buf[STITCH_INPUT_NODE]);
+		if (addrs[4] == 0 ||
+		    !(hw_enables[0] & PISP_BE_BAYER_ENABLE_STITCH_INPUT) ||
+		    !(hw_enables[0] & PISP_BE_BAYER_ENABLE_STITCH)) {
+			hw_enables[0] &=
+				~(PISP_BE_BAYER_ENABLE_STITCH_INPUT |
+				  PISP_BE_BAYER_ENABLE_STITCH_DECOMPRESS |
+				  PISP_BE_BAYER_ENABLE_STITCH);
+		}
+
+		addrs[5] = pispbe_get_addr(buf[TDN_OUTPUT_NODE]);
+		if (addrs[5] == 0)
+			hw_enables[0] &= ~(PISP_BE_BAYER_ENABLE_TDN_COMPRESS |
+					   PISP_BE_BAYER_ENABLE_TDN_OUTPUT);
+
+		addrs[6] = pispbe_get_addr(buf[STITCH_OUTPUT_NODE]);
+		if (addrs[6] == 0)
+			hw_enables[0] &=
+				~(PISP_BE_BAYER_ENABLE_STITCH_COMPRESS |
+				  PISP_BE_BAYER_ENABLE_STITCH_OUTPUT);
+	} else {
+		/* No Bayer input? Disable entire Bayer pipe (else lockup) */
+		hw_enables[0] = 0;
+	}
+
+	/* Main image output channels. */
+	for (unsigned int i = 0; i < PISP_BACK_END_NUM_OUTPUTS; i++) {
+		ret = pispbe_get_planes_addr(addrs + 7 + 3 * i,
+					     buf[OUTPUT0_NODE + i],
+					     &node_group->node[OUTPUT0_NODE + i]);
+		if (ret <= 0)
+			hw_enables[1] &= ~(PISP_BE_RGB_ENABLE_OUTPUT0 << i);
+	}
+
+	/* HoG output (always single plane). */
+	addrs[13] = pispbe_get_addr(buf[HOG_OUTPUT_NODE]);
+	if (addrs[13] == 0)
+		hw_enables[1] &= ~PISP_BE_RGB_ENABLE_HOG;
+}
+
+/*
+ * Prepare a job description to be submitted to the HW.
+ *
+ * To schedule a job, we need all streaming nodes (apart from Output0,
+ * Output1, Tdn and Stitch) to have a buffer ready, which must
+ * include at least a config buffer and a main input image.
+ *
+ * For Output0, Output1, Tdn and Stitch, a buffer only needs to be
+ * available if the blocks are enabled in the config.
+ *
+ * Needs to be called with hw_lock held.
+ *
+ * Returns 0 if a job has been successfully prepared, < 0 otherwise.
+ */
+static int pispbe_prepare_job(struct pispbe_node_group *node_group,
+			      struct pispbe_job_descriptor *job)
+{
+	struct pispbe_dev *pispbe = node_group->pispbe;
+	struct pispbe_buffer *buf[PISPBE_NUM_NODES];
+	unsigned int config_index;
+	struct pispbe_node *node;
+	unsigned long flags;
+
+	lockdep_assert_held(&pispbe->hw_lock);
+
+	memset(job, 0, sizeof(struct pispbe_job_descriptor));
+
+	if (((BIT(CONFIG_NODE) | BIT(MAIN_INPUT_NODE)) &
+		node_group->streaming_map) !=
+			(BIT(CONFIG_NODE) | BIT(MAIN_INPUT_NODE)))
+		return -ENODEV;
+
+	node = &node_group->node[CONFIG_NODE];
+	spin_lock_irqsave(&node->ready_lock, flags);
+	buf[CONFIG_NODE] = list_first_entry_or_null(&node->ready_queue,
+						    struct pispbe_buffer,
+						    ready_list);
+	if (buf[CONFIG_NODE])
+		list_del(&buf[CONFIG_NODE]->ready_list);
```

Comment:
What happens to this buffer now if `prepare_job()` subsequently fails because other nodes/buffers were not queued by the user? Is it lost forever?

Reply:
Hi @njhollinghurst. Can we please move the comments to the ML, otherwise keeping this and the posted version in sync is difficult. 

Anyway, I guess the same question applies to all buffers which are extracted from a queue,  for each node. 
In the for loop that iterates over the video device we have

```
		/* Pull a buffer from each V4L2 queue to form the queued job */
		spin_lock_irqsave(&node->ready_lock, flags);
		buf[i] = list_first_entry_or_null(&node->ready_queue,
						  struct pispbe_buffer,
						  ready_list);
		if (buf[i]) {
			list_del(&buf[i]->ready_list);
			pispbe->queued_job.buf[i] = buf[i];
		}
		spin_unlock_irqrestore(&node->ready_lock, flags);

		if (!buf[i] && !ignore_buffers)
			return -ENODEV;
```

if we error out earlier with -ENODEV all buf[i] which has been extracted from a queue should probably be queued back ?

Reply:
Happy to revert `205735176035 ("be: Delete buffers from queue in a single loop")` if we have to push more fixes on this version. I'm not too bothered about the two loops, as said it should mostly be harmless. The worst that can happen is that a queue that resulted empty at inspection time (inside the first for loop) receives a new buffer in between the two loops and gets ignored. The other only conflicting path is a `stream_stop`  while we're preparing the job, which returns all buffers from the queue in between these two for loops.

Reply:
I'd find it helpful to discuss here just a little more, so I don't say something misguided on the ML.

Can you explain the concern about receiving a new buffer in between the two loops? What is the difference, in terms of locking and atomicity, between the two code versions?

Reply:
> I'd find it helpful to discuss here just a little more, so I don't say something misguided on the ML.
> 

no worries

> Can you explain the concern about receiving a new buffer in between the two loops? What is the difference, in terms of locking and atomicity, between the two code versions?
>

The `ready_queue` queue is accessed in three locations
1) When preparing a job, to extract buffers from its front
2) When queuing a buffer to its tail
3) When stopping streaming to clear the queue

in 1) the queue is accessed in two points, one to inspect it if there's any available buffer, and at a second time to actually remove buffers from the queue. There's then a tiny window where the queue might be accessed by 2) and 3) before actually being modified.

As said, it should be harmless, but if in example the one of the nodes has no buffers queued at the time of inspection, then a buffers gets queued, it will get ignored and could potentially cause a mis-alignement.

In general, inspecting and modifying an object which is accessed by other concurrent code paths should be done while holding a lock, to reduce potentially hard to debug issues. If the solution turns out to be more complex than the original issue was then we can evaluate if it's worth it.

> Is there a better design in which we can push the removed buffers back onto the _front_ of the queues?

Isn't pushing in front a simple `list_add()` ?

Reply:
I think I understand. So removing the buffers and (conditionally) adding them again on failure, seems like a clear improvement over the old two-loops design, but is still theoretically vulnerable to an intervening `STREAMOFF`. It might be good enough.

If we acquire all the queue locks (in a safe order!?), could it be possible to do the whole operation atomically?

Reply:
> I think I understand. So removing the buffers and (conditionally) adding them again on failure, seems like a clear improvement over the old two-loops design, but is still theoretically vulnerable to an intervening `STREAMOFF`. It might be good enough.
> 

`stop_streaming` is actually protected by `hw_lock`, and `pispbe_prepare_job` is run with the same lock held, so should be safe in that regard now that I look more closely.

So it's only about making sure `pispbe_prepare_job()` and `queue_buffer` do not step on each other. I think returning all buffers to the queue front in case of error is probably the safest option.

> If we acquire all the queue locks (in a safe order!?), could it be possible to do the whole operation atomically?

Do you mean  `pispbe_prepare_job()` with "the whole operation" ?

Reply:
> Do you mean `pispbe_prepare_job()` with "the whole operation" ?

I did (at least, all of the queue inspection and state changes within that function). But it sounds like that is not necessary.

Reply:
@njhollinghurst I've re-pushed the pull request adding on top https://github.com/raspberrypi/linux/pull/5899/commits/9fcfb8ee070ecc2ca9b4fcf4629c4545f9cf178e

I triggered the error and tested with cam, which hangs as no job gets scheduled. I wonder if in the case of errors in creating the job description wouldn't be better to stop the streaming operations completely..

Reply:
Thanks - I'll try to do some testing this afternoon.

From a first look at the code, are all of `buf[]` initialized when we get to `err_return_buffers`?

Reply:
Should they ? Only the initialized ones are relevant, as they are initialized with buffers extracted from the ready_list queue. If  `buf[i]` has not been initialized it means no buffer have been extracted from the queue. Did I get your question correctly ?

Reply:
If we terminate the loop with `goto`, it looks like some value of `buf[i]` will contain an undefined value.

`buf[]` might need to be initialized before the loop, or upon termination, or maybe the cleanup could be done inside the loop (without a goto) so it can see the current value of the loop counter?

Reply:
ah right, `struct pispbe_buffer *buf[PISPBE_NUM_NODES];` is not initialized.

From memory
`pispbe_buffer *buf[PISPBE_NUM_NODES] = {}; `

should be valid C99, otherwise we can easily memset() ?

Reply:
Yes, that seems to work!

One way to test this stuff is to modify `src/libcamera/pipeline/rpi/pisp/pisp.cpp:prepareBe()`, to move the line
`        isp_[Isp::Input].queueBuffer(buffer)`
to the bottom of the function, after the `Config` buffer has been queued. (This is not normal user behaviour.)

Without initializing `buf[]`, that would result in a hard crash, dereferencing a bad virtual address in `pispbe_prepare_job()`. With the empty initializer, it appears to be fixed.